File: Study Plan/LeetCode 75/1207 Unique Number of Occurrences/UniqueNumberOfOccurrencesV3.java
/*
    Problem
    --------------------------------------------------------------------------------
    1207. Unique Number of Occurrences
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/unique-number-of-occurrences/solutions/4555311/java-based-on-hints-beats-98-in-runtime-time-and-space-complexity-o-n/
    
    - Runtime 2 ms (Beats 98.13% of users with Java)
    - Memory 41.56 MB (Beats 31.99% of users with Java)
*/

class Solution {
    public boolean uniqueOccurrences(int[] arr) {
        // Hint 1
        // Find the number of occurrences of each element in the array
        // using a hash map.
        HashMap<Integer, Integer> occurrences = new HashMap<>();
        for (int element : arr) {
            int count = occurrences.getOrDefault(element, 0) + 1;
            occurrences.put(element, count);
        }
        HashSet<Integer> uniques = new HashSet<>();
        // Hint 2
        // Iterate through the hash map and check if there is a repeated value.
        for (int occurrence : occurrences.values()) {
            // Returns true if this set did not already contain the specified element
            // https://docs.oracle.com/javase/8/docs/api/java/util/HashSet.html#add-E-
            if(!uniques.add(occurrence))
            {
                return false;
            }
        }
        return true;
    }
}

File: Study Plan/LeetCode 75/1207 Unique Number of Occurrences/UniqueNumberOfOccurrencesV3.cs
/*
    Problem
    --------------------------------------------------------------------------------
    1207. Unique Number of Occurrences
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/unique-number-of-occurrences/solutions/4643870/c-based-on-hints-beats-95-in-runtime-time-and-space-complexity-o-n/

    - Runtime 61 ms (Beats 95.05% of users with C#)
    - Memory 41.85 MB (Beats 89.02% of users with C#)
*/

public class Solution
{
    public bool UniqueOccurrences(int[] arr)
    {
        // Hint 1
        // Find the number of occurrences of each element in the array
        // using a hash map.
        var occurrences = new Dictionary<int, int>(arr.Length);

        foreach (var element in arr)
        {
            if (occurrences.ContainsKey(element))
            {
                occurrences[element]++;
            }
            else
            {
                occurrences[element] = 1;
            }
        }

        var uniques = new HashSet<int>();

        // Hint 2
        // Iterate through the hash map and check if there is a repeated value.
        foreach (var occurrence in occurrences.Values)
        {
            // Returns true if the element is added to the HashSet<T> object,
            // false if the element is already present.
            // https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add
            if(!uniques.Add(occurrence))
            {
                return false;
            }
        }
        
        return true;
    }
}